### ISSUE TITLE
RichText in markdown mode gets confused when wrapping and a line ends with code span

### ISSUE THREAD

**Issue by Goltanju** [fyne-io/fyne#4303]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same problem, before opening a new one.
- [X] This issue only relates to a single bug. I will open new issues for any other problems.

### Describe the bug

I've noticed that when set to fyne.TextWrapWord and a line ends with a code span, the widget gets very confused and wraps at 1 character onward.

### How to reproduce

Some example text I was working with was:

"This is a test of the RichText widget's interactions between markdown and `word` wrapping."

And then just play around with the display width of the widget and at some point you should see it look like:

```
This is a test of the RichText widget's interactions between markdown and word
w
r
a
p
p
i
n
g
.
```

With the example code, just run it an then slowly change the window width and observe the odd behaviour.

Also, side-note that I could create a whole new Issue for, but... Anyway, why in this sample program is the window so hugely tall?

### Screenshots

![image](https://github.com/fyne-io/fyne/assets/137658871/201c88bf-fb0f-437b-b9a2-d9562e3963af)

### Example code

```go
package main

import (
    "fyne.io/fyne/v2"
    "fyne.io/fyne/v2/app"
    "fyne.io/fyne/v2/container"
    "fyne.io/fyne/v2/widget"
)

func main() {
    app.NewWithID("go.source.20231020")
    win := fyne.CurrentApp().NewWindow("test20231020")
    richText := widget.NewRichTextFromMarkdown("This is a test of the RichText widget's interactions between markdown   and `word` wrapping.")
    richText.Wrapping = fyne.TextWrapWord
    win.SetContent(container.NewMax(richText))
    win.ShowAndRun()
}
```

### Fyne version

2.4.1

### Go compiler version

1.21.0

### Operating system and version

Windows 11

**Comment by Goltanju** [comment]
Okay I updated that first comment with example code and screenshot. It's no longer an issue for me as it turns out my app doesn't need Markdown after all, but I figured I'd leave the report since it'll probably affect _somebody_.

**Comment by theHilikus** [comment]
Any updates on this bug? I see the repro code uses a deprecated function NewMax
I am able to reproduce both bugs (weird single char wrapping and super tall window) with this slightly modified code as well

```go
package main

import (
	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/app"
	"fyne.io/fyne/v2/widget"
)

func main() {
	a := app.New()
	w := a.NewWindow("Hello World")

	markdown := widget.NewRichTextFromMarkdown("Hello World with **bold** text that is very long and will wrap because of that. [a link](https://example.com) " +
		"Loooooooooooooong word after")
	markdown.Wrapping = fyne.TextWrapWord
	w.SetContent(markdown)
	w.Resize(fyne.NewSize(400, 600))
	w.ShowAndRun()
}
```

If anyone can provide any pointers I might give it a try to fix it

In case it matters (don't think so) I have seen this bug in Linux and Windows

<img width="300" height="1408" alt="Image" src="https://github.com/user-attachments/assets/11abc5ba-8ce3-4f59-ab49-8f07c4d00378" />

**Comment by redawl** [comment]
I think part of the bug here is that `TextWrapWord` starts to behave like `TextWrapBreak` once the windows is shrunk to smaller than one of the words. You can see this with this example code if you shrink the window horizontally:
```go
package main

import (
        "fmt"

        "fyne.io/fyne/v2"
        "fyne.io/fyne/v2/app"
        "fyne.io/fyne/v2/test"
        "fyne.io/fyne/v2/widget"
)

func main() {
        a := app.New()
        w := a.NewWindow("Hello World")

        markdown := widget.NewLabel("Hello World with text that is very long")
        markdown.Wrapping = fyne.TextWrapWord
        w.SetContent(markdown)
        fmt.Println(test.RenderToMarkup(w.Canvas()))
        w.Resize(fyne.NewSize(400, 600))
        w.ShowAndRun()
}
```

**Comment by andydotxyz** [comment]
You're right - it's confusing having no room at all with having no room left on this line

**Comment by redawl** [comment]
IMO `TextWrapWord` should not allow the label to wrap on a char level, even if it can only fit a single word. It should report a MinSize that doesn't allow the window to shrink smaller than the largest word. What do you think @andydotxyz

**Comment by andydotxyz** [comment]
I'm happy to agree with that. I cannot remember a specific reason it was implemented to degrade like that.

**Comment by theHilikus** [comment]
I tried to fix this issue. As much as what @redawl said makes sense to me about the minsize problem, it looks to me like the problem here is even before that. If we fixed only the problem with the MinSize, the wrapping would still behave incorrectly where every word (instead of every character) would be in its own line.

From what I can tell, the problem is that we don't detect the end of line when the space left is consumed by a hyperlink segment

**Comment by MaxGyver83** [comment]
Btw, this issue is not limited to code spans and hyperlinks. I get this even for plain text:

<img width="498" height="119" alt="Image" src="https://github.com/user-attachments/assets/e5ff2493-d76e-461b-a5ee-685ddf682903" />

```go
text := `Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque x.
Mauris`
```

`n.Dump(source, 0)`:

```
Document {
    Paragraph {
        RawText: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque x.
Mauris"
        HasBlankPreviousLines: true
        Text(SoftLineBreak): "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque x."
        Text: "Mauris"
    }
}
```

**Comment by dweymouth** [comment]
Oops, closed too soon since #6200 only partially fixes this, reopening...

**Comment by MaxGyver83** [comment]
> Oops, closed too soon since [#6200](https://github.com/fyne-io/fyne/pull/6200) only partially fixes this, reopening...

I have finished #6201 and rebased it to `develop`. Now, it's ready for a review.

**Comment by andydotxyz** [comment]
Excellent thanks

**Comment by ErikKalkoken** [comment]
@MaxGyver83 Thank you so much for fixing this bug. This was a blocker for me for using Richtext with Links in my apps. I am very glad that I can finally use it.

**Comment by theHilikus** [comment]
any idea when this will be released?

**Comment by andydotxyz** [comment]
The main change releases today now that the vote on v2.7.4 has passed.
Some bug fixes in RichText had to be delayed to v2.8.0 as the codebases had diverged too far. The release process for that begins soon and it should be out in June.
